server/src/test/java/org/elasticsearch/cluster/IndexingWithMappingUpdateIT.java

```diff
@@ -0,0 +1,193 @@
+        final CountDownLatch finishLatch = new CountDownLatch(1 + docs);
+
+        AtomicInteger bulkMessagesCounter = new AtomicInteger();
+
+        MockTransportService replicaTransportService =
+            ((MockTransportService) cluster.getInstance(TransportService.class, replicaNodeName));
+
+        // tracer on replica to slow down consumption of some requests :
+        // internal:discovery/zen/publish/sent , internal:discovery/zen/publish/commit - bring cluster state with mapping update
+        // indices:data/write/bulk - indexing request
+        replicaTransportService.addTracer(new MockTransportService.Tracer() {
+            @Override
+            public void receivedRequest(long requestId, String action) {
+                boolean clusterStateCommit = action.equals(PublishClusterStateAction.COMMIT_ACTION_NAME);
+                boolean clusterStateWrite = action.equals(PublishClusterStateAction.SEND_ACTION_NAME);
+                boolean firstBulkRequestSend = bulkMessagesCounter.getAndIncrement() == 0 &&
+                    action.startsWith(BulkAction.NAME);
+
+                if (clusterStateCommit) {
+                    sendSecondDocLatch.countDown();
+                }
+                if (clusterStateCommit || clusterStateWrite || firstBulkRequestSend) {
+                    try {
+                        Thread.sleep(50);
+                    } catch (InterruptedException e) {
+                        // ignore
+                    }
+                }
+                super.receivedRequest(requestId, action);
+            }
+        });
+
+        final List<IndexResponse> indexResponses = new CopyOnWriteArrayList<>();
+        new Thread(() -> {
+            try {
+                prestartLatch.countDown();
+                startLatch.await();
+
+                IndexResponse indexResponse = client.prepareIndex(index, type, "doc")
+                    .setSource("{ \"f\": \"normal\"}", XContentType.JSON).get();
+                indexResponses.add(indexResponse);
+            } catch (InterruptedException e) {
+                // ignore
+            } finally {
+                finishLatch.countDown();
+            }
+        }).start();
+
+        for(int i = 0; i < docs; i++) {
+            IndexRequestBuilder builder = client.prepareIndex(index, type, "doc_" + i)
+                .setSource("{ \"f\": \"normal\"}", XContentType.JSON);
+            new Thread(() -> {
+                try {
+                    prestartLatch.countDown();
+                    sendSecondDocLatch.await();
+
+                    IndexResponse indexResponse = builder.get();
+                    indexResponses.add(indexResponse);
+                } catch (InterruptedException e) {
+                    // ignore
+                } finally {
+                    finishLatch.countDown();
+                }
+            }).start();
+        }
+
+        prestartLatch.await();
+        startLatch.countDown();
+        finishLatch.await();
```

Comment:
Could we just call `.join()` on each thread we just created to guarantee that they're all finished? Using the `finishLatch` like this doesn't technically guarantee that the threads are all done by the time the test finishes, although in practice I'm sure it's fine.